File: src/python/multithreading/01_single_worker_thread.py
"""
This script demonstrates the use of threads in Python.
It shows how to create a worker thread, retrieve its ID, and print the process ID.

When the script is executed, it will display the process ID and thread ID for both the main function
and the worker function. It will also display the thread names, which can be useful for debugging
and monitoring purposes.
"""

import os
import threading
from threading import Thread


def foo():
    print("Worker function")
    print(f"Worker function process id: {os.getpid()}")
    print(f"Worker thread id: {threading.currentThread().ident}")
    print(f"Worker thread name: {threading.currentThread().getName()}")


if __name__ == "__main__":

    print(f"Main function process id: {os.getpid()}")
    print(f"Main thread id: {threading.currentThread().ident}")
    print(f"Main thread name: {threading.currentThread().getName()}")

    thread = Thread(name="Example_thread", target=foo)
    thread.start()

    thread.join()

File: src/cpp/multithreading/01_single_worker_thread.cpp
#include <iostream>
#include <string>
#include <sys/syscall.h>
#include <thread>
#include <unistd.h>

void foo() {
  std::cout << "Worker function" << std::endl;
  std::cout << "Worker function process id: " << getpid() << std::endl;
  std::cout << "Worker thread id: " << std::this_thread::get_id() << std::endl;
  std::cout << "Worker thread name: Example_thread" << std::endl;
}

int main() {
  std::cout << "Main function process id: " << getpid() << std::endl;
  std::cout << "Main thread id: " << std::this_thread::get_id() << std::endl;
  std::cout << "Main thread name: main" << std::endl;

  std::thread worker_thread(foo);

  // if we use join main thread will wait for that thread to finish
  worker_thread.join();

  // other option is to detach the thread and then it will proceede freely on
  // it's own --daemon process worker_thread.detach();

  // you can join and detach only once

  std::cout << std::boolalpha << worker_thread.joinable() << std::endl;

  return 0;
}
